examples/sssp/sssp.cu

```diff
@@ -1,44 +1,73 @@
 #include <cstdlib>  // EXIT_SUCCESS
 
-#include <gunrock/applications/sssp/sssp.hxx>
+#include <gunrock/applications/runner.hxx>
+#include <gunrock/applications/sssp/sssp_implementation.hxx>
 
 using namespace gunrock;
 using namespace memory;
 
 void test_sssp(int num_arguments, char** argument_array) {
-  using vertex_t = int;
-  using edge_t = int;
-  using weight_t = float;
-
+  
   if (num_arguments != 2) {
-    std::cerr << "usage: ./bin/color filename.mtx" << std::endl;
+    std::cerr << "usage: ./bin/<program-name> filename.mtx" << std::endl;
     exit(1);
   }
-
-  // Load Matrix-Market file & convert the resultant COO into CSR format.
+  
+  // --
+  // Define types
+  
+  using vertex_t = int;
+  using edge_t   = int;
+  using weight_t = float;
+  
+  // --
+  // IO
+  
   std::string filename = argument_array[1];
+  
   io::matrix_market_t<vertex_t, edge_t, weight_t> mm;
-  auto coo = mm.load(filename);
-
-  // convert coo to csr
-  format::csr_t<memory_space_t::device, vertex_t, edge_t, weight_t> csr;
-  csr = coo;  // Able to convert host-based coo_t to device-based csr (or host
-              // to host). As of right now, it requires coo to be host side.
+  format::csr_t<memory::memory_space_t::device, vertex_t, edge_t, weight_t> csr;
+  csr.from_coo(mm.load(filename)); 
 
-  vertex_t source = 0;
-  thrust::device_vector<weight_t> d_distances(csr.number_of_rows);
+  // --
+  // Build graph + metadata
+  
+  auto [G, meta] = graph::build::from_csr_t<memory_space_t::device>(&csr);
+  
+  using graph_t = decltype(G)::value_type;
+  using meta_t  = decltype(meta)::value_type;
+  
+  // --
+  // Params and memory allocation
+  
+  vertex_t single_source = 0;
+  
+  vertex_t n_vertices = meta[0].get_number_of_vertices();
+  thrust::device_vector<weight_t> distances(n_vertices);
+  thrust::device_vector<vertex_t> predecessors(n_vertices);
+  
+  // --
+  // Setup problem
+  
+  using param_t   = sssp::sssp_param_t<meta_t>;
+  using result_t  = sssp::sssp_result_t<meta_t>;
+  using problem_t = sssp::sssp_problem_t<graph_t, meta_t>;
+  using enactor_t = sssp::sssp_enactor_t<problem_t>;
```

Comment:
John mentioned maybe we should find a way to hide all these, but as of right now I am not sure how to do that.

Reply:
This is updated in https://github.com/bkj/essentials/pull/4